src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -254,3 +255,505 @@ g.test('inputs')
+        maxComputeWorkgroupSizeZ < t.params.sizes[2],
+      'Workgroup size too large'
+    );
+
+    const wgsl = `
+enable subgroups;
+
+const stride = ${wgThreads};
+
+@group(0) @binding(0)
+var<storage, read_write> output : array<u32>;
+
+@group(0) @binding(1)
+var<storage, read_write> compare : array<u32>;
+
+@compute @workgroup_size(${t.params.sizes[0]}, ${t.params.sizes[1]}, ${t.params.sizes[2]})
+fn main(@builtin(subgroup_size) size : u32,
+        @builtin(workgroup_id) wgid : vec3u,
+        @builtin(local_invocation_index) lid : u32) {
+  output[lid + wgid.x * stride] = size;
+  let ballot = countOneBits(subgroupBallot(true));
+  let ballotSize = ballot[0] + ballot[1] + ballot[2] + ballot[3];
+  compare[lid + wgid.x * stride] = ballotSize;
+}`;
+
+    const numInvocations = wgThreads * t.params.numWGs;
+    const sizesBuffer = t.makeBufferWithContents(
+      new Uint32Array([...iterRange(numInvocations, x => 0)]),
+      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST
+    );
+    t.trackForCleanup(sizesBuffer);
+    const compareBuffer = t.makeBufferWithContents(
+      new Uint32Array([...iterRange(numInvocations, x => 0)]),
+      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST
+    );
+    t.trackForCleanup(compareBuffer);
+
+    const pipeline = t.device.createComputePipeline({
+      layout: 'auto',
+      compute: {
+        module: t.device.createShaderModule({
+          code: wgsl,
+        }),
+        entryPoint: 'main',
+      },
+    });
+    const bg = t.device.createBindGroup({
+      layout: pipeline.getBindGroupLayout(0),
+      entries: [
+        {
+          binding: 0,
+          resource: {
+            buffer: sizesBuffer,
+          },
+        },
+        {
+          binding: 1,
+          resource: {
+            buffer: compareBuffer,
+          },
+        },
+      ],
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bg);
+    pass.dispatchWorkgroups(t.params.numWGs, 1, 1);
+    pass.end();
+    t.queue.submit([encoder.finish()]);
+
+    const sizesReadback = await t.readGPUBufferRangeTyped(sizesBuffer, {
+      srcByteOffset: 0,
+      type: Uint32Array,
+      typedLength: numInvocations,
+      method: 'copy',
+    });
+    const sizesData: Uint32Array = sizesReadback.data;
+
+    const compareReadback = await t.readGPUBufferRangeTyped(compareBuffer, {
+      srcByteOffset: 0,
+      type: Uint32Array,
+      typedLength: numInvocations,
+      method: 'copy',
+    });
+    const compareData: Uint32Array = compareReadback.data;
+
+    t.expectOK(checkSubgroupSizeConsistency(sizesData, compareData, minSize, maxSize, wgThreads));
+  });
+
+/**
+ * Checks the consistency of subgroup_invocation_id builtin values
+ *
+ * @param data The subgroup_invocation_id output data
+ * @param
```

Comment:
Unfinished comment?

Reply:
Yes, this needs better comments.

>  @param ids:  output of 'subgroup_id's storage buffer.  ids[gid] should be the global id of the first member of the subgroup containing the invocation with global id gid.

Or if you want to clean that up you can make a definition:  
>  The *leader* of  subgroup is the invocation in the subgroup with the lowest global id.
> Then ids[gid] is the global id of the leader of the subgroup containing gid.